core/matrix/csr.cpp

```diff
@@ -618,6 +621,92 @@ void Csr<ValueType, IndexType>::write(mat_data& data) const
 }
 
 
+template <typename ValueType, typename IndexType, typename TransformClosure>
+std::pair<std::unique_ptr<Csr<ValueType, IndexType>>,
+          typename Csr<ValueType, IndexType>::permuting_reuse_info>
+transform_reusable(const Csr<ValueType, IndexType>* input, gko::dim<2> out_size,
+                   size_type nnz, TransformClosure closure)
+{
+    using FloatIndexType =
+        std::conditional_t<std::is_same_v<IndexType, int32>, float, double>;
+    static_assert(sizeof(FloatIndexType) == sizeof(IndexType));
+    static_assert(alignof(FloatIndexType) == alignof(IndexType));
+    auto exec = input->get_executor();
+    auto in_size = input->get_size();
+    auto transformed = Csr<ValueType, IndexType>::create(exec, out_size, nnz);
+    // transform matrix with integer values from 0 to nnz - 1 reinterpret_cast
+    // as float
+    array<IndexType> iota_values{exec, nnz};
+    exec->run(csr::make_fill_seq_array(iota_values.get_data(), nnz));
+    auto iota_float_view = make_array_view(
+        exec, nnz, reinterpret_cast<FloatIndexType*>(iota_values.get_data()));
+    auto iota_mtx = Csr<FloatIndexType, IndexType>::create_const(
+        exec, input->get_size(), iota_float_view.as_const_view(),
+        make_const_array_view(exec, nnz, input->get_const_col_idxs()),
+        make_const_array_view(exec, in_size[0] + 1,
+                              input->get_const_row_ptrs()),
+        std::make_shared<typename Csr<FloatIndexType, IndexType>::sparselib>());
+    auto transformed_iota = closure(iota_mtx.get());
+    exec->copy(out_size[0] + 1, transformed_iota->get_const_row_ptrs(),
+               transformed->get_row_ptrs());
+    exec->copy(nnz, transformed_iota->get_const_col_idxs(),
+               transformed->get_col_idxs());
+    exec->copy(nnz,
+               reinterpret_cast<const IndexType*>(
+                   transformed_iota->get_const_values()),
+               iota_values.get_data());
+    auto transform_permutation =
+        Permutation<IndexType>::create(exec, std::move(iota_values));
+    transformed->set_strategy(input->get_strategy());
+    // permute values into output matrix
+    input->create_const_value_view()->permute(transform_permutation,
+                                              transformed->create_value_view(),
+                                              permute_mode::rows);
+
+    return std::make_pair(
+        std::move(transformed),
+        typename Csr<ValueType, IndexType>::permuting_reuse_info{
+            std::move(transform_permutation)});
+}
+
+
+template <typename ValueType, typename IndexType>
+Csr<ValueType, IndexType>::permuting_reuse_info::permuting_reuse_info()
+    : permuting_reuse_info{nullptr}
```

Comment:
Isn't this just the default constructor?

Reply:
this is calling the other permuting_reuse_info constructor, not the value_permutation constructor, if that's what you mean?

Reply:
Oh, I see. But doesn't that still have the same effect as the default constructor?

Reply:
yes, I just wanted to base everything on a single implementation of the constructor, no deep other reason behind it.